### PR TITLE
fix(js): make module and shared-memory ownership worker-safe

### PR DESCRIPTION
**Pull request by syrusakbary**
# Description

Move cross-worker Module and SharedMemory handling into Wasmer's JavaScript backend, without changing WASIX's dynamic linker state or TaskWasm interfaces.

The browser can structured-clone compiled WebAssembly modules and shared memories, but wasm-bindgen references belong to one worker. Shared Rust ownership now uses stable IDs; JavaScript references stay in worker-local registries. Cloning/dropping portable Rust wrappers never accesses another worker's JS handle table.

This is a separate backend-level alternative to #6977, related to #4158. The older PR is unchanged. The SDK companion prototype removes 471 net lines of special transport/reconstruction and unused notification code; SDK changes and test-only raw-pointer probes are not included here.

## Shared memory only

Following maintainer review, the detached byte-copy representation has been removed entirely. JS VMSharedMemory contains only a shared-object handle and MemoryType, with a fallible conversion checking both the type flag and actual SharedArrayBuffer backing.

- `as_shared()` produces a handle to the same shared memory.
- `copy()` produces an independent copy which is also shared memory.
- **JS behavior change:** non-shared `copy()` now returns `MemoryNotShared`, matching the existing sys and V8 behavior. `copy_to_store()` currently delegates to this path and likewise rejects non-shared memory.
- Ordinary non-shared Memory allocation/access remains supported. No MemorySnapshot type or non-shared copy API is introduced in this PR.
- Public documentation now states the shared-only contract rather than promising non-shared copies through SharedMemory.

This replaces the earlier prototype's non-shared snapshot support and its intermediate-allocation optimization. No WASIX, TaskWasm, dynamic-linker, dependency, or native-backend behavior changes are needed.

## Transport and lifetime

- Two stateless hooks: `prepare_shared_object_message(payload)` attaches all live local objects; `receive_shared_object_message(message)` validates/imports them before returning the payload. There are no transport objects, per-connection known-ID sets, prepared-envelope types, acknowledgements, or send-commit calls.
- Every task-bearing message carries a full snapshot. Retrying a failed send or sending to a replacement worker uses the same protocol, without resetting delivery state. The SDK prototype attaches snapshots only when boxing a Rust task; lifecycle/control messages stay plain.
- Receivers validate envelope namespace, entry shapes, integer IDs, duplicates, and object kinds before installing entries. Expired IDs cannot resurrect dead owners. This is an unsafe trusted-runtime protocol: validation does not authenticate object identity.
- Cleanup is opportunistic. Registries track Rust ownership with `Weak<Owner>`; registration and snapshot boundaries release entries whose owners have expired. Hosts can also call `collect_shared_objects()` at task boundaries.
- No timers, BroadcastChannel, microtask notifications, or `Owner::drop` hook. The cleanup helper file has been deleted. Idle workers can retain expired JS references until their next collection boundary or teardown; there is intentionally no eventual-idle-cleanup guarantee.
- The JS values remain strongly rooted while registered: substituting JS `WeakRef` would let GC invalidate a still-live Rust module or detached shared-memory handle. Rust weak-owner tracking does not itself keep those Rust owners alive.
- The SDK needs no per-object cleanup routing. Removing background cleanup subtracts another 91 net lines after the earlier stateless snapshot simplification.

## Module-delivery guarantee

Successful `receive_shared_object_message` installs the envelope's live objects **before returning the task payload**. Tests include a module without retained bytes, so fallback compilation cannot hide missing delivery at that boundary.

This cannot cover modules published later through shared Rust linker state while a sibling is executing synchronously. The browser regression queues delivery before waking a blocked worker, but that worker still reaches module access before its message handler runs. The retained-bytes fallback compiles once locally and remains necessary for this case. Normally transported modules do not recompile.

An unconditional delivery-before-access guarantee would require cooperative suspension/publication barriers across linker operations, not merely another message/acknowledgement. See the [worker event-loop model](https://html.spec.whatwg.org/multipage/workers.html#the-event-loop). Missing-module/memory errors identify the transport hooks and required import-before-access ordering.

## Validation

- **11 JS-backend unit tests passed**, executed with wasm-bindgen-test-runner in Node: weak Rust ownership and explicit collection after final drop, snapshot validation/lifetime, repeated full snapshots, import-before-dispatch without retained bytes, non-shared rejection (including mismatched JS/type metadata), and independent shared copies.
- **28 native API library tests and all 7 memory integration tests passed**, including non-shared copy/copy_to_store rejection.
- **Actual in-app browser integration passed** on the SDK-compatible port (the SDK pins a different Wasmer base):
  - shared writes/growth, independent shared copies, non-shared detachment rejection, nested transfer/final drop, and 25 cleanup cycles;
  - repeated full snapshots, real DataCloneError/retry without bookkeeping, replacement-worker delivery, plain lifecycle messages, idle references retained then released at the next snapshot, and registration/drop without opening cleanup channels or timers;
  - modules without retained bytes, dispatch-before-access, and the blocked-worker late-module counterexample;
  - Python **3.13.18** threads, subprocess threads, Rich progress, files/streams/cache, and late dlopen with an already-running sibling.
- Cranelift CLI build, JS library/test Clippy, and full `make lint` passed, including Taplo.
- SDK production wasm build, size check (4.51 MB raw / 1.16 MB Brotli), and TypeScript build passed. Worker integration probes live in the separate SDK harness, not the upstream unit suite.
- A small browser round-trip smoke covers 0 / 16 / 64 attached objects (50 messages after warmup). This includes snapshot export/import and worker messaging, is not a comparative benchmark, and does not establish sustained-load performance. An earlier larger allocation batch hit `WebAssembly.Memory(): could not allocate memory` before transport; it is not counted as passing coverage.

## Review focus

Ready for human API/lifetime review. Remaining explicit constraints:

- Every task snapshot scans and resends all live local objects, not task-reachable objects only. This deliberately trades repeated structured-clone overhead for a smaller stateless protocol; lifecycle messages do not pay that cost.
- Idle-worker retention is an explicit policy tradeoff: expired objects remain rooted until collection or worker teardown, regardless of when JS GC runs.
- Shared memory and modules without retained bytes must arrive before synchronous access. Fallback compilation requires retained bytes and does not promise preservation of arbitrary custom compile options.
- Idle worker termination/replacement was tested. Force-terminating workers holding Rust locks/ownership is not made safe by this PR.
- Sustained concurrency, cancellation, and broader browser coverage remain follow-up test areas.
- The separate Python subprocess cwd limitation is not addressed; this is not a claim that all pip installations succeed.

Implemented and submitted by Codex, an OpenAI coding agent, at the repository owner's request.

### CODE REVIEW

File: lib/api/src/backend/js/entities/memory/mod.rs [hunk 1]
```diff
@@ -200,7 +200,7 @@ impl Memory {
 
     pub fn copy(&self, _store: &impl AsStoreRef) -> Result<SharedMemory, MemoryError> {
         Ok(SharedMemory::new(crate::vm::VMSharedMemory::Js(
-            self.handle.copy()?,
+            self.handle.copy()?.into(),
         )))
     }
```
[the remark this answers is not in the document]
Comment (syrusakbary) [reply lib/api/src/backend/js/entities/memory/mod.rs:208]: Addressed in 4ce0cab1fd9. Non-shared Memory::copy now snapshots the source handle directly; only shared copies allocate a separate WebAssembly.Memory before detachment. Added backend tests for grown-memory contents and independent attachment, plus a browser constructor-count test proving that source creation + non-shared copy performs exactly one WebAssembly.Memory allocation. Shared-memory copies remain independent. — Codex

File: lib/api/src/backend/js/entities/module.rs [hunk 2]
```diff
@@ -88,20 +88,29 @@ fn evaluate_i32_init_expr(
...
+        // A running sibling can observe dlopen before receiving postMessage.
+        #[cfg(feature = "js-serializable-module")]
+        if let Some(bytes) = &self.raw_bytes {
+            let bytes = Uint8Array::from(bytes.as_ref());
+            let module = WebAssembly::Module::new(&bytes.into())
+                .expect("failed to compile a shared module in this worker");
+            self.module.install(module.clone());
+            return module;
+        }
+        panic!("module was not transported to this worker and has no retained bytes");
```
[the remark this answers is not in the document]
Comment (syrusakbary) [reply lib/api/src/backend/js/entities/module.rs]: Addressed in 4ce0cab1fd9. The panic now points to SharedObjectTransport/receive_shared_object_message, the export_shared_objects/import_shared_objects alternative, and the feature + retained-bytes requirement for fallback compilation. Added an import-before-dispatch test using a module without retained bytes. The PR description also documents the tested delivery boundary and why queued delivery cannot cover late synchronous sibling access without cooperative suspension. — Codex

File: lib/api/src/backend/js/vm/memory.rs [hunk 3]
```diff
@@ -104,5 +104,48 @@ impl From<VMMemory> for (JsValue, MemoryType) {
...
+        }
+    }
+}
+impl VMSharedMemory {
+    pub fn attach(self) -> VMMemory {
+        let memory = match self.memory {
+            DetachedMemory::Shared(handle) => handle
+                .get()
+                .expect("shared memory was not transported to this worker"),
+            DetachedMemory::Copied(bytes) => {
```
[the remark this answers is not in the document]
Comment (syrusakbary) [reply lib/api/src/backend/js/vm/memory.rs]: Addressed in 4ce0cab1fd9. The missing-memory panic names SharedObjectTransport/receive_shared_object_message and the export/import alternative, explicitly requiring import before attachment. Copied-memory allocation failures now format the underlying MemoryError explicitly. The existing expect already included a debug representation; the new message makes the actionable error text explicit. — Codex

File: lib/api/src/backend/js/entities/memory/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (syrusakbary) [reply lib/api/src/backend/js/entities/memory/mod.rs:208]: Following the maintainer’s requested simplification, 1df51a0ac71 removes the non-shared detached-copy path entirely rather than retaining this optimization. SharedMemory now represents only shared memory; non-shared copy() and the delegating copy_to_store() return MemoryNotShared, matching existing sys/V8 behavior. The PR description explicitly records this JS behavior change. Non-shared detachment rejection and independent shared-copy behavior pass backend/native/browser tests. No separate snapshot API is introduced here. — Codex

File: lib/api/src/backend/js/entities/module.rs [hunk 4]
```diff
@@ -88,20 +88,34 @@ fn evaluate_i32_init_expr(
...
 impl Module {
+    fn local_module(&self) -> WebAssembly::Module {
+        if let Some(module) = self.module.get() {
+            return module;
+        }
+        // A running sibling can observe dlopen before receiving postMessage.
+        #[cfg(feature = "js-serializable-module")]
+        if let Some(bytes) = &self.raw_bytes {
+            let bytes = Uint8Array::from(bytes.as_ref());
+            let module = WebAssembly::Module::new(&bytes.into())
```
Comment (Arshia001) [inline lib/api/src/backend/js/entities/module.rs:106]: Chrome and Firefox forbid synchronous `new WebAssembly.Module` on the main browser thread for buffers over 4 KB, so this `expect` aborts whenever a module compiled in a worker is first touched on the main thread — any `exports()` call, e.g. from the WASIX linker or `Instance::new`.

`from_binary_unchecked` has the same constraint but returns `Result<_, CompileError>`. Here there's no recovery path, and `local_module()` also converts previously-infallible accessors (`exports`, `imports`, `custom_sections`, `From<Module> for JsValue`) into panics.

File: lib/api/src/backend/js/vm/memory.rs [hunk 5]
```diff
@@ -104,5 +104,78 @@ impl From<VMMemory> for (JsValue, MemoryType) {
...
+        }
+        Ok(Self {
+            memory: crate::js::utils::shared_handle::SharedJsHandle::new(memory.memory.into_inner()),
+            ty: memory.ty,
+        })
+    }
+}
+impl VMSharedMemory {
+    pub fn attach(self) -> VMMemory {
+        let memory = self.memory.get().expect(
```
Comment (Arshia001) [inline lib/api/src/backend/js/vm/memory.rs:131]: This `expect` is reachable from a fallible path: `SpawnType::AttachMemory(mem) => Ok(Some(mem.attach(store)))` in `lib/wasix/src/runtime/task_manager/mod.rs:209`, which otherwise returns `Result<_, WasiThreadError>`.

A host that hasn't wired the new `prepare_shared_object_message` / `receive_shared_object_message` hooks — the SDK companion change isn't in this PR — gets a hard panic in the spawned worker instead of a recoverable `MemoryCreateFailed`. Can `attach`/`into_vm_memory` return `Result` instead?

File: lib/api/src/backend/js/entities/memory/mod.rs [hunk 6]
```diff
@@ -199,22 +199,21 @@ impl Memory {
...
 
     pub fn copy(&self, _store: &impl AsStoreRef) -> Result<SharedMemory, MemoryError> {
+        if !self.handle.ty.shared {
```
Comment (Arshia001) [inline lib/api/src/backend/js/entities/memory/mod.rs:202]: This regresses `proc_fork` on the JS backend: `lib/wasix/src/syscalls/wasix/proc_fork.rs:183` calls `SpawnType::AttachMemory(memory.copy(...))` and traps on `Err`. Previously the byte-for-byte `VMMemory::copy()` served this, so a single-threaded (non-shared-memory) WASIX binary could `fork()` in the browser; it now traps with `MemoryNotShared`.

The API tightening looks intentional, but this consequence isn't mentioned and there's no compensating path.

File: lib/api/src/backend/js/utils/shared_handle.rs [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+    owner: Arc<Owner>,
+}
+impl PartialEq for SharedJsHandle {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+impl Eq for SharedJsHandle {}
+impl SharedJsHandle {
+    pub fn new(value: impl Into<JsValue>) -> Self {
```
Comment (Arshia001) [inline lib/api/src/backend/js/utils/shared_handle.rs:68]: Every construction runs a full `collect_shared_objects()` plus an O(n) linear `JsValue`-equality scan over all registered objects. With dynamic linking this is called once per loaded module, so it's quadratic in module count.

File: lib/api/src/backend/js/utils/shared_handle.rs [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+pub fn collect_shared_objects() {
+    OBJECTS.with_borrow_mut(|objects| objects.retain(|_, object| object.owner.strong_count() > 0));
+    OWNERS
+        .lock()
+        .unwrap()
+        .retain(|_, owner| owner.strong_count() > 0);
+}
+
+/// Export all live local objects for structured cloning within one trusted runtime.
+pub fn export_shared_objects() -> Array {
```
Comment (Arshia001) [inline lib/api/src/backend/js/utils/shared_handle.rs:131]: This structured-clones *every* live `WebAssembly.Module` in the worker, and `prepare_shared_object_message` calls it for every task-bearing message. Under WASIX dynamic linking (dozens to hundreds of `.so` modules) each thread spawn re-serializes every compiled module — browsers serialize compiled code for `WebAssembly.Module`, so this is tens of MB per `postMessage`, not a constant factor.

The PR notes structured-clone overhead as a tradeoff, but the smoke test (0/16/64 objects) doesn't reach the dynamic-linking case.

File: lib/api/src/backend/js/utils/shared_handle.rs [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+                id.is_finite() && *id >= 1.0 && *id <= f64::from(u32::MAX) && id.fract() == 0.0
+            })
+            .ok_or_else(|| JsValue::from_str("invalid shared-object ID"))? as u32;
+        if !seen.insert(id) {
+            return Err(JsValue::from_str("duplicate shared-object ID"));
+        }
+        let value = entry.get(1);
+        let kind = ObjectKind::of(&value)
+            .ok_or_else(|| JsValue::from_str("invalid shared-object type"))?;
+        let owner = OWNERS.lock().unwrap().get(&id).and_then(Weak::upgrade);
```
Comment (Arshia001) [inline lib/api/src/backend/js/utils/shared_handle.rs:172]: The lock is acquired and released inside the per-entry loop, so an N-entry snapshot takes the global mutex N times.

On wasm32 with atomics a contended `std::sync::Mutex` blocks via `memory.atomic.wait32`, which traps on the browser main thread — e.g. the main thread calling `Memory::as_shared()` or compiling a `Module` while a worker is mid-`import_shared_objects`. Hoisting the lock out of the loop fixes the N acquisitions and shrinks the window.

File: lib/api/src/backend/js/utils/shared_handle.rs [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+            if owner.kind != kind {
+                return Err(JsValue::from_str("shared-object type mismatch"));
+            }
+            entries.push((id, value, owner));
+        }
+    }
+    collect_shared_objects();
+    OBJECTS.with_borrow_mut(|objects| {
+        for (id, value, owner) in &entries {
+            objects.entry(*id).or_insert_with(|| LocalObject {
```
Comment (Arshia001) [inline lib/api/src/backend/js/utils/shared_handle.rs:183]: `or_insert_with` never replaces an existing local entry. After a worker takes the `local_module()` recompile fallback and installs under id N, a later authoritative snapshot for id N is silently ignored — and that worker's own `export_shared_objects()` then propagates its recompiled object to other workers.

So `JsValue::from(module)` can yield different JS objects per worker for the same Rust `Module`, and the recompiled copy's loss of the original compile options spreads rather than staying local.

File: lib/api/src/backend/js/utils/shared_handle.rs [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+///
+/// # Safety
+/// The envelope must be produced by this runtime's prepare_shared_object_message on the
+/// same trusted worker connection; see import_shared_objects.
+pub unsafe fn receive_shared_object_message(message: JsValue) -> Result<JsValue, JsValue> {
+    if !Array::is_array(&message) {
+        return Ok(message);
+    }
+    let envelope: Array = message.unchecked_into();
+    if envelope.length() != 3 || envelope.get(0).as_string().as_deref() != Some(NAMESPACE.as_str())
```
Comment (Arshia001) [inline lib/api/src/backend/js/utils/shared_handle.rs:226]: Any `Array` message that isn't an envelope errors here, which contradicts the doc above ("Plain non-array lifecycle messages pass through unchanged") — `[cmd, arg]` tuples are a common worker convention and are now rejected.

The error path also drops `envelope.get(1)`, so a host whose payload is a leaked `Box` pointer leaks the box and permanently pins every `Arc<Owner>` it holds; those JS objects can then never be collected in any worker. A version/kind tag check that falls through to pass-through would be safer than erroring on every array.

**Review by Arshia001 — CHANGES_REQUESTED** [review]
The ID + `Arc<Owner>`/`Weak` ownership design is sound — I traced the drop/collect/import ordering, `Weak::upgrade` resurrection, ID monotonicity, and the `OWNERS`/`OBJECTS` lock nesting, and found no use-after-free or deadlock. Removing `unsafe impl Send/Sync for Module` is safe, and moving the shared check into `TryFrom<VMMemory> for VMSharedMemory` is a strict improvement over the old metadata-only guard.

Requesting changes on the panic surfaces and the two behavior regressions below — specifically (1), which aborts on the browser main thread, and (3), which breaks `proc_fork` on this backend.

One more, not inline because the file isn't in the diff: `BackendMemory::copy` / `copy_to_store` in `lib/api/src/entities/memory/inner.rs:173` and `:215` still carry the old "detached copied memory handle" doc text. The public `Memory::copy` doc was updated to the shared-only contract; these two now describe behavior no backend has.
